ocaml/xapi/xapi_cluster.ml

```diff
@@ -47,6 +47,50 @@ let create ~__context ~pIF ~cluster_stack ~pool_auto_join ~token_timeout ~token_
       assert_pif_prerequisites (pIF,pifrec);
       let ip = ip_of_pif (pIF,pifrec) in
 
+      let assert_pool_on_same_subnet () =
+        (** we gather ip addresses from all the hosts and also the netmask from the PIF.
+          * if all the ip addresses do not have the same network address, then we raise an
+            error *)
+        let are_on_same_subnet =
+            let parse_ip ip_str =
+              match Ipaddr.of_string ip_str with
+              | Ok ip_addr       -> ip_addr
+              (* this shouldn't happen because we shouldn't be storing invalid ips *)
+              | Error (`Msg msg) -> raise Api_errors.(Server_error (internal_error, [Printf.sprintf "could not parse ip address (%s), msg: %s" ip_str msg]))
+            in
+            let pool_host_ips =
+               Xapi_pool_helpers.get_master_slaves_list ~__context |>
+               List.map (fun x -> Db.Host.get_address ~__context ~self:x) |>
+               List.map parse_ip
+            in
+            let ensure_same_protocol ips =
+              let netmask = pifrec.API.pIF_netmask |> parse_ip in
+              let coerce_v4 xs = xs |> List.map (function
+                | Ipaddr.V4 x -> x
+                | V6 _        -> raise Api_errors.(Server_error (internal_error, ["expected IPv4, but got IPv6"])))
+              in
+              let coerce_v6 xs = xs |> List.map (function
+                | Ipaddr.V6 x -> x
+                | V4 _        -> raise Api_errors.(Server_error (internal_error, ["expected IPv6, but got IPv4"])))
+              in
+              match netmask with
+              | Ipaddr.V4 nm -> `Ip_v4s (nm, coerce_v4 ips)
+              | V6        nm -> `Ip_v6s (nm, coerce_v6 ips)
+            in
+            match ensure_same_protocol pool_host_ips with
+            | `Ip_v4s (netmask, ip::ips) ->
```

Comment:
All this special casing and coercing of v4/v6 feel like something the ipaddr library should be able to abstract and hide from us...?

Reply:
i mentioned above that the joint interface isn't as powerful (as far as i can tell).

with the joint v4/v6 interface you can't provide a complete ip address as a netmask, only using as a single integer; and I assume we have to support arbitrary netmasks (at least in the ipv4 case)?